video/vdu_sprites.h

```diff
@@ -175,41 +198,63 @@ void VDUStreamProcessor::vdu_sys_sprites(void) {
 	}
 }
 
-void VDUStreamProcessor::receiveBitmap(uint8_t cmd, uint16_t width, uint16_t height) {
-	auto bufferId = getCurrentBitmapId();
-
+void VDUStreamProcessor::receiveBitmap(uint16_t bufferId, uint16_t width, uint16_t height) {
 	// clear the buffer
 	bufferClear(bufferId);
+	// receive the data
+	if (bufferWrite(bufferId, sizeof(uint32_t) * width * height) != 0) {
+		// timed out, or couldn't allocate buffer - so abort
+		return;
+	}
+	// create RGBA8888 bitmap from buffer
+	createBitmapFromBuffer(bufferId, 0, width, height);
+}
+
+void VDUStreamProcessor::createBitmapFromScreen(uint16_t bufferId) {
+	bufferClear(bufferId);
+	// get screen rectangle from last two graphics cursor positions
+	auto x1 = p1.X;
+	auto y1 = p1.Y;
+	auto x2 = p2.X;
+	auto y2 = p2.Y;
+	auto width = x2 - x1;
+	auto height = y2 - y1;
+	// normalize to positive values
+	if (width < 0) {
+		width = -width;
+		x1 = x2;
+	}
+	if (height < 0) {
+		height = -height;
+		y1 = y2;
+	}
+	auto size = width * height;
+	if (size == 0) {
+		debug_log("vdu_sys_sprites: bitmap %d - zero size\n\r", bufferId);
+		return;
+	}
+
+	// create a new buffer of appropriate size, and set as a native format bitmap
+	auto buffer = bufferCreate(bufferId, size);
+	createBitmapFromBuffer(bufferId, 3, width, height);
+	// Copy screen area to buffer
+	canvas->copyToBitmap(x1, y1, getBitmap(bufferId).get());
```

Comment:
this copyToBitmap routine is new - code for it can be seen in AgonConsole8/vdp-gl#4

this command is added to the canvas drawing processing queue.  this has a few effects to bear in mind

the primary reason to do this is to ensure that the capture will reflect all drawing operations up to the point that the bitmap is created.

as this is treated as part of the drawing queue, sprites will be hidden and not included in the area of canvas copied to the bitmap.

a down-side of this approach is that the buffer won't get updated to contain the bitmap data until after the drawing queue has been processed.  buffer commands to process the bitmap data (such as commands to copy or mirror the data) therefore won't work until after that has happened.  the canvas drawing queue can be explicitly flushed by waiting for vsync.

the bitmap is, however, available to be drawn immediately after this call is made.  i.e. you can send a command to copy an area of screen to a bitmap, and then immediately draw that bitmap elsewhere on the screen.  this works because the bitmap data will be present when the bitmap drawing command is processed by the canvas drawing queue.

this function _could_ potentially perform a queue flush just before calling the `copyToBitmap`, however doing so may cause unnecessary sprite flickering